## python 기본 library인 functools 에 있는 partial 이라는 method를 알아보자  
partial은 function에 여러 args들이 요구될 때, 몇개는 미리 값을 넣어두고 변화하는 하나의 값만을 사용하여 Function call을 하는 것을 도와준다.  
무슨소리인지 모르겠지만 아래 code를 보면 이해 가능  

만드려고 하는 함수는 덧셈을 무조건 어떤 값 n 에다가 해서 값을 내보내는 function이다.  
굳이 이런 것을 왜 하나 싶지만 그냥 이해를 위한 단순한 function이기 때문이고, 활용만 잘한다면 엄청나게 깔끔한 코드 및 노가다를 방지할 수 있다.  

In [3]:
def add_mul(mode, *args):
    if mode == 'add':
        result = 0
        for i in args:
            result += i
        return result
    
    elif mode == 'mul':
        result = 1
        for i in args:
            result = result * i
        return result
    else:
        raise ValueError("wrong mode input. only 'add' or 'mul'")

In [7]:
add_value = add_mul('add', 1,2,3,4)
add_value

10

기본적으로 사용하는 식으로는 위처럼 사용한다. 1,2,3,4가 function의 args로 넘어가면서 list 형식으로 넘어간다.  
args = list of values [1,2,3,4]  
이런 식으로 넘어간 것!  

function에 무조건 어떤 값을 넘겨놓고, 추가로 어떤 값을 넘기면서 call 하고 싶을 땐 어떨까?  
예를 들어, lr_scheduler에서 기본 max_iter나 max_epoch이나 변하지 않는 값은 function을 선언할 때 미리 넣어두고,  
현재 iter만을 넘겨서 lr_scheduling을 하고 싶은 경우가 있다.  
그런거 미리 function에 선언해두면 되지 않냐 싶지만 experiment가 계속 변경되면서 max_epoch도 바뀌고 그럴텐데, 그럴 때마다 함수 찾아들어가서 바꿀건가 ..?  
  
우선은 위의 코드로 미리 넣어두고 추가의 값들을 넘기면서 어떤 식으로 되는지 보자

In [9]:
from functools import partial
add_mul_function = partial(add_mul, 'add', 100)
add_value = add_mul_function(1,2,3,4)

add_value

110

add_mul_function 이라는 Function을 partial 을 이용해서 기존의 add_mul function에 미리 add와 100이라는 값을 주고,  
그 함수에 1,2,3,4 라는 값을 넘겨서 110이라는 값을 주었다.  
'add', 100은 add_mul_function을 사용할 때마다 항상 미리 들어가 있는 값들이고, 거기에 내가 추가로 1, 2, 3, 4 라는 값을 주는 것!  